### ISSUE TITLE
Binary masks instead of indices?

### ISSUE THREAD

**Issue by aydoo** [rusty1s/pytorch_scatter#177]
Using indices e.g. in the index tensor of the `scatter` function limits it to mapping values to exactly _one_ reduction node. Using binary masks would generalize this to arbitrary mappings. You can of course rewrite the index tensor to a 'one-hot-encoding' style the way it currently works to emulate the support for binary masks, but this leads to a lot of unnecessary calculations for where the mask is 0. In a context where this binary mask is very sparse, this is not feasible. It would be great if there was an out-of-the-box way to deal with arbitrary mappings.

**Comment by rusty1s** [comment]
I'm not sure if I understand you correctly. Can you show me a simple example to understand this use-case?

**Comment by aydoo** [comment]
```
import torch
from torch_scatter import scatter-

masks = torch.tensor([[0,1,1,0,0],[0,1,0,1,0]]).T # [5,2], i.e. two 5D binary masks
source = torch.arange(5) # [5], 5D source

source = source.expand(masks.shape[1],-1).T # [5,2], expanding because scatter doesn't auto broadcast source

sums = scatter(source, masks, dim=0, reduce='sum')
sums_where_0 = sums[0] # sums where each mask is zero -- this is unnecessarily calculated)
sums_where_1 = sums[1] # sums where each mask is one -- this is what we want

print(source)
# tensor([[0, 0],
#         [1, 1],
#         [2, 2],
#         [3, 3],
#         [4, 4]])

print(masks)
# tensor([[0, 0],
#         [1, 1],
#         [1, 0],
#         [0, 1],
#         [0, 0]])

print(sums_where_1)
# tensor([3, 4])
```
Note that `sums_where_1 = sums[1]` gives me the sums where masks are `1`, but this only works because both masks start with a `0` entry, otherwise the order in sums is different.

**Comment by rusty1s** [comment]
I see. I don't think the current approach is that bad. `sums_where_1 = sums[1]` should always work (even if masks do not start with a `0`). A slightly faster approach would be to do the following:
```python
x = mask.to(x.device) * x
x.sum(dim=0)
```
